### PR TITLE
Add some evaluation timings

### PR DESCRIPTION
**Pull request by MorinoseiMorizo**
I added some evaluation timings, which include the number of samples, epochs, target words and processing time.

### CODE REVIEW

File: src/bin/train.cc [hunk 1]
```diff
@@ -395,6 +396,7 @@ int main(int argc, char * argv[]) {
         config.get<unsigned>("Batch.batch_size"),
         train_max_length, train_max_length_ratio,
         config.get<unsigned>("Global.random_seed"));
+    const int corpus_size = train_sampler.getCorpusSize();
```
Comment (odashi) [inline src/bin/train.cc]: const unsigned.

Comment (odashi) [inline src/bin/train.cc]: const unsigned.

File: src/bin/train.cc [hunk 2]
```diff
@@ -406,12 +408,18 @@ int main(int argc, char * argv[]) {
...
+        % train_sampler.getCorpusSize() % dev_sampler.getCorpusSize() 
+        % test_sampler.getCorpusSize();
+    logger->info(fmt_corpus_size.str());
     nmtkit::BatchConverter batch_converter(*src_vocab, *trg_vocab);
 
     dynet::Model model;
 
     // Creates a new trainer.
     boost::scoped_ptr<dynet::Trainer> trainer(::createTrainer(config, &model));
+    trainer->sparse_updates_enabled = false;
```
Comment (odashi) [inline src/bin/train.cc]: Some comments for this line are recommended. (I guess this line should be required for correct updating when using some trainers with momentum)

File: src/bin/train.cc [hunk 3]
```diff
@@ -432,12 +440,28 @@ int main(int argc, char * argv[]) {
     float lr_decay = 1.0f;
     const float lr_decay_ratio = config.get<float>("Train.lr_decay_ratio");
 
-    // Train/dev/test loop
     const float dropout_ratio = config.get<float>("Train.dropout_ratio");
     const unsigned max_iteration = config.get<unsigned>("Train.max_iteration");
+
+    const string evaluation_type = config.get<string>("Train.evaluation_type");
```
Comment (odashi) [inline src/bin/train.cc]: Unify names "evaluation" and "eval."

File: src/bin/train.cc [hunk 4]
```diff
@@ -432,12 +440,28 @@ int main(int argc, char * argv[]) {
...
+    unsigned long next_eval_samples = eval_interval;
+    if (evaluation_type == "epoch") {
+      next_eval_samples = eval_interval * corpus_size;
+    } else if (evaluation_type == "sample") {
+      next_eval_samples = eval_interval;
+    }
+    auto next_eval_time = 
+        std::chrono::system_clock::to_time_t(
+        (std::chrono::system_clock::now() + std::chrono::minutes(eval_interval)));
+    auto training_start_time = std::chrono::system_clock::now();
```
Comment (odashi) [inline src/bin/train.cc]: const?

File: src/bin/train.cc [hunk 5]
```diff
@@ -432,12 +440,28 @@ int main(int argc, char * argv[]) {
...
     unsigned long num_trained_samples = 0;
+    unsigned long num_trained_words = 0;
+    unsigned long next_eval_words = eval_interval;
+    unsigned long next_eval_samples = eval_interval;
+    if (evaluation_type == "epoch") {
+      next_eval_samples = eval_interval * corpus_size;
+    } else if (evaluation_type == "sample") {
+      next_eval_samples = eval_interval;
+    }
+    auto next_eval_time =
```
Comment (odashi) [inline src/bin/train.cc]: Multiple function call system_clock::now() is redundant and may generate different timestamp for each variable. Just call it only once.

File: src/bin/train.cc [hunk 6]
```diff
@@ -456,49 +480,74 @@ int main(int argc, char * argv[]) {
         trainer->update(lr_decay);
 
         num_trained_samples += batch.source_ids[0].size();
+        num_trained_words += batch.target_ids.size() * batch.target_ids[0].size();
```
Comment (odashi) [inline src/bin/train.cc]: Some comments are required because this calculation also includes the number of padding words.

File: src/bin/train.cc [hunk 7]
```diff
@@ -456,49 +480,74 @@ int main(int argc, char * argv[]) {
...
         if (!train_sampler.hasSamples()) {
           train_sampler.rewind();
         }
 
-        const string fmt_str = "Trained: batch=%d samples=%d lr=%.6e";
+        const string fmt_str = "Trained: batch=%d samples=%d current-proc-words=%d lr=%.6e";
```
Comment (odashi) [inline src/bin/train.cc]: current-proc-words -> words (maybe short name is better because of reducing log file size)

File: src/bin/train.cc [hunk 8]
```diff
@@ -456,49 +480,74 @@ int main(int argc, char * argv[]) {
...
+            % iteration % num_trained_samples % num_trained_words % lr_decay;
         logger->info(fmt.str());
       }
 
       if (lr_decay_type == "batch") {
         lr_decay *= lr_decay_ratio;
       }
 
-      if (iteration % eval_interval == 0) {
+      if (num_trained_samples >= next_epoch_samples) {
```
Comment (odashi) [inline src/bin/train.cc]: This condition looks equal to !train_sampler.hasSamples() and below lines could be replaced to above if-statement (line 484).

File: src/bin/train.cc [hunk 9]
```diff
@@ -456,49 +480,74 @@ int main(int argc, char * argv[]) {
...
+        auto eval_start_time = std::chrono::system_clock::now();
+
         logger->info("Evaluating...");
 
         const float dev_log_ppl = ::evaluateLogPerplexity(
             encdec, dev_sampler, batch_converter);
         const auto fmt_dev_log_ppl = boost::format(
-            "Evaluated: batch=%d dev-log-ppl=%.6e")
-            % iteration % dev_log_ppl;
+            "Evaluated: batch=%d current-proc-words=%d elapsed-time(sec)=%d dev-log-ppl=%.6e")
```
Comment (odashi) [inline src/bin/train.cc]: current-proc-words -> words

File: src/bin/train.cc [hunk 10]
```diff
@@ -456,49 +480,74 @@ int main(int argc, char * argv[]) {
...
+            % iteration % num_trained_words % elapsed_time_seconds % dev_log_ppl;
         logger->info(fmt_dev_log_ppl.str());
 
         const float dev_bleu = ::evaluateBLEU(
             *trg_vocab, encdec, dev_sampler, train_max_length);
         const auto fmt_dev_bleu = boost::format(
-            "Evaluated: batch=%d dev-bleu=%.6f")
-            % iteration % dev_bleu;
+            "Evaluated: batch=%d current-proc-words=%d elapsed-time(sec)=%d dev-bleu=%.6f")
```
Comment (odashi) [inline src/bin/train.cc]: current-proc-words -> words

File: src/bin/train.cc [hunk 11]
```diff
@@ -456,49 +480,74 @@ int main(int argc, char * argv[]) {
...
+            % iteration % num_trained_words % elapsed_time_seconds % dev_bleu;
         logger->info(fmt_dev_bleu.str());
 
         const float test_log_ppl = ::evaluateLogPerplexity(
             encdec, test_sampler, batch_converter);
         const auto fmt_test_log_ppl = boost::format(
-            "Evaluated: batch=%d test-log-ppl=%.6e")
-            % iteration % test_log_ppl;
+            "Evaluated: batch=%d current-proc-words=%d elapsed-time(sec)=%d test-log-ppl=%.6e")
```
Comment (odashi) [inline src/bin/train.cc]: current-proc-words -> words

File: src/bin/train.cc [hunk 12]
```diff
@@ -456,49 +480,74 @@ int main(int argc, char * argv[]) {
...
+            % iteration % num_trained_words % elapsed_time_seconds % test_log_ppl;
         logger->info(fmt_test_log_ppl.str());
 
         const float test_bleu = ::evaluateBLEU(
             *trg_vocab, encdec, test_sampler, train_max_length);
         const auto fmt_test_bleu = boost::format(
-            "Evaluated: batch=%d test-bleu=%.6f")
-            % iteration % test_bleu;
+            "Evaluated: batch=%d current-proc-words=%d elapsed-time(sec)=%d test-bleu=%.6f")
```
Comment (odashi) [inline src/bin/train.cc]: ditto

File: sample_data/sample_config.ini [hunk 13]
```diff
@@ -180,6 +180,13 @@ dropout_ratio=0.3
 ; Maximum number of batch data to be trained.
 max_iteration=10000
 
+; Timing of evaluating dev/test set. Available options:
+; * step .... Number of steps (iterations).
+; * sample .... Number of samples (sentences).
+; * epoch .... Number of epochs.
```
Comment (odashi) [inline sample_data/sample_config.ini]: "epoch" is ambiguous. It might be better to change "corpus."

File: src/include/nmtkit/sampler.h [hunk 14]
```diff
@@ -26,6 +26,11 @@ class Sampler {
   //   List of new samples.
   virtual std::vector<Sample> getSamples() = 0;
 
+  // Rerieves the size of the corpus.
+  // Returns:
+  //   Size of the corpus.
```
Comment (odashi) [inline src/include/nmtkit/sampler.h]: This function might returns the number of "filtered" samples. It should be noted in the comment.

File: src/include/nmtkit/sampler.h [hunk 15]
```diff
@@ -26,6 +26,11 @@ class Sampler {
   //   List of new samples.
   virtual std::vector<Sample> getSamples() = 0;
 
+  // Rerieves the size of the corpus.
```
Comment (odashi) [inline src/include/nmtkit/sampler.h]: "corpus size" is ambiguous. "number of samples (or sentences)" might be better.

File: src/include/nmtkit/sampler.h [hunk 16]
```diff
@@ -26,6 +26,11 @@ class Sampler {
...
+  // Returns:
+  //   Size of the corpus.
+  virtual int getCorpusSize() = 0;
```
Comment (odashi) [inline src/include/nmtkit/sampler.h]: getCorpusSize -> getNumSamples (or other good name)

File: src/bin/train.cc [hunk 17]
```diff
@@ -456,49 +480,74 @@ int main(int argc, char * argv[]) {
...
+        next_epoch_samples += corpus_size;
+        auto elapsed_time = std::chrono::system_clock::now() - epoch_start_time;
+        auto elapsed_time_seconds = std::chrono::duration_cast<std::chrono::seconds>(elapsed_time).count();
+        const auto fmt_epoch_time = boost::format(
+                "Epoch finished: elapsed-time(sec)=%d") % elapsed_time_seconds;
+        logger->info(fmt_epoch_time.str());
+        epoch_start_time = std::chrono::system_clock::now();
+      }
+
+      if ((evaluation_type == "step" and iteration % eval_interval == 0) or
```
Comment (odashi) [inline src/bin/train.cc]: Highly complex condition. it should be separated out the if-statement.

File: src/bin/train.cc (hunk 4 above)
Comment (MorinoseiMorizo) [inline src/bin/train.cc]: training_start_time should be const, but next_eval_time will be changed later.

**Comment by MorinoseiMorizo** [comment]
Thank you for the review and comments.
I fixed the codes based on your comments.
Can you review it, please?

Thank you in advance.

File: sample_data/sample_config.ini [hunk 18]
```diff
@@ -180,6 +180,13 @@ dropout_ratio=0.3
...
+; Timing of evaluating dev/test set. Available options:
+; * step .... Number of steps (iterations).
+; * sample .... Number of samples (sentences).
+; * corpus .... Number of epochs (corpus).
+; * word .... Number of processed target words.
+; * time .... Processing time (minutes).
+evaluation_type=step
 ; Evaluation (validation/testing) processes would be performed in each following
-; step.
+; step/sample/epoch/word/minute.
```
Comment (odashi) [inline sample_data/sample_config.ini]: Fix comment.

File: src/bin/train.cc [hunk 19]
```diff
@@ -406,12 +408,25 @@ int main(int argc, char * argv[]) {
...
+        % train_sampler.getNumSamples() % dev_sampler.getNumSamples() 
+        % test_sampler.getNumSamples();
+    logger->info(fmt_corpus_size.str());
     nmtkit::BatchConverter batch_converter(*src_vocab, *trg_vocab);
 
     dynet::Model model;
 
     // Creates a new trainer.
     boost::scoped_ptr<dynet::Trainer> trainer(::createTrainer(config, &model));
+    // Disable sparse updates.
```
Comment (odashi) [inline src/bin/train.cc]: Recommend to add header to clarify responsibility of comments, like:
// NOTE(MorinoseiMorizo): blah blah blah ...

File: src/include/nmtkit/sampler.h [hunk 20]
```diff
@@ -26,6 +26,11 @@ class Sampler {
   //   List of new samples.
   virtual std::vector<Sample> getSamples() = 0;
 
+  // Rerieves the number of filtered samples.
+  // Returns:
+  //   Number of filtered samples.
+  virtual int getNumSamples() = 0;
```
Comment (odashi) [inline src/include/nmtkit/sampler.h]: unsigned.

File: src/bin/train.cc [hunk 21]
```diff
@@ -456,49 +488,73 @@ int main(int argc, char * argv[]) {
         trainer->update(lr_decay);
 
         num_trained_samples += batch.source_ids[0].size();
+        num_trained_words +=  (batch.target_ids.size() - 1) * batch.target_ids[0].size();
         if (!train_sampler.hasSamples()) {
           train_sampler.rewind();
+          auto elapsed_time = std::chrono::system_clock::now() - epoch_start_time;
```
Comment (odashi) [inline src/bin/train.cc]: const.

File: src/bin/train.cc [hunk 22]
```diff
@@ -456,49 +488,73 @@ int main(int argc, char * argv[]) {
...
+          auto elapsed_time_seconds = std::chrono::duration_cast<std::chrono::seconds>(elapsed_time).count();
```
Comment (odashi) [inline src/bin/train.cc]: const.

**Comment by MorinoseiMorizo** [comment]
Thank you for the review.
I just fixed it.
